lib/improver/ensemble_calibration.py

```diff
@@ -0,0 +1,1418 @@
+            warnings.warn(msg)
+            return optimised_coeffs, coeff_names
+
+        rename_coordinate(
+            current_forecast_cubes, "ensemble_member_id", "realization")
+        rename_coordinate(
+            historic_forecast_cubes, "ensemble_member_id", "realization")
+
+        current_forecast_cubes = concatenate_cubes(
+            current_forecast_cubes)
+        historic_forecast_cubes = concatenate_cubes(
+            historic_forecast_cubes)
+        truth_cubes = concatenate_cubes(truth_cubes)
+
+        for current_forecast_cube in current_forecast_cubes.slices_over(
+                "time"):
+            date = unit.num2date(
+                current_forecast_cube.coord("time").points,
+                current_forecast_cube.coord("time").units.name,
+                current_forecast_cube.coord("time").units.calendar)[0]
+            # Extract desired forecast_period from historic_forecast_cubes.
+            forecast_period_constr = iris.Constraint(
+                forecast_period=current_forecast_cube.coord(
+                    "forecast_period").points)
+            historic_forecast_cube = historic_forecast_cubes.extract(
+                forecast_period_constr)
+
+            # Extract truth matching the time of the historic forecast.
+            truth_constr = iris.Constraint(
+                forecast_reference_time=historic_forecast_cube.coord(
+                    "time").points)
+            truth_cube = truth_cubes.extract(truth_constr)
+
+            if truth_cube is None:
+                msg = ("Unable to calibrate for the time points {} "
+                       "as no truth data is available."
+                       "Moving on to try to calibrate "
+                       "next time point.".format(
+                           historic_forecast_cube.coord("time").points))
+                warnings.warn(msg)
+                continue
+
+            # Make sure inputs have the same units.
+            historic_forecast_cube.convert_units(self.desired_units)
+            truth_cube.convert_units(self.desired_units)
+
+            if self.predictor_of_mean_flag.lower() in ["mean"]:
+                no_of_members = None
+                forecast_predictor = historic_forecast_cube.collapsed(
+                    "realization", iris.analysis.MEAN)
+            elif self.predictor_of_mean_flag.lower() in ["members"]:
+                no_of_members = len(
+                    historic_forecast_cube.coord("realization").points)
+                forecast_predictor = historic_forecast_cube
+
+            forecast_var = historic_forecast_cube.collapsed(
+                "realization", iris.analysis.VARIANCE)
+
+            # Computing initial guess for EMOS coefficients
+            # If no initial guess from a previous iteration, or if there
+            # are NaNs in the initial guess, calculate an initial guess.
+            if "initial_guess" not in locals() or nan_in_initial_guess:
```

Comment:
Yes, the first time around the loop over time slices, `nan_in_initial_guess` will be False, however, it could be set to True in further iterations, if there are NaN values present in the calculated value for the `initial_guess`.